### ISSUE TITLE
problem with R-4.0

### ISSUE THREAD

**Issue by tdhock** [r-hub/rhub#310]
Hi I am getting this error message in 00install.out,
```
* installing *source* package 'nc' ...
** using staged installation
** R
** inst
** byte-compile and prepare package for lazy loading
Error: .onLoad failed in loadNamespace() for 'data.table', details:
  call: fun(libname, pkgname)
  error: This is R 4.0.0 but data.table has been installed using R 3.6.1. The major version must match. Please reinstall data.table.
Execution halted
ERROR: lazy loading failed for package 'nc'
* removing 'C:/Users/USERtiZzbmglJA/R/nc'
```

this is for build https://builder.r-hub.io/status/nc_2019.10.19.tar.gz-afcd91dfbbb044c78fe1c4cc85382593

**Comment by gaborcsardi** [comment]
Well, I believe that in this case I think data.table is wrong, and 4.0 is compatible with 3.6 builds.

**Comment by tdhock** [comment]
Hi @gaborcsardi thanks for the quick response!

@jangorecki @mattdowle do you know how to fix this problem with using data.table on R-hub? (which seems to be building data.table under R-3.6.1 but running it under R-4.0.0)

**Comment by gaborcsardi** [comment]
R-hub does not build data.table at all, just uses the CRAN binary from https://cran.rstudio.com/bin/windows/contrib/3.7/data.table_1.12.6.zip which is built on R 3.6.1, and is still compatible with 4.0.0

**Comment by jangorecki** [comment]
@tdhock Thanks for pinging. I filled https://github.com/Rdatatable/data.table/issues/4009

**Comment by mattdowle** [comment]
Please could you reopen. data.table is not wrong. data.table 1.12.6+ does need to be reinstalled in R4+ before it will work in R 4+. The binaries are indeed compatible as you said: the same `.zip` can be installed that was built using R3. But it is when the .zip is being installed that the `if()` in NAMESPACE is executed by the version of R being used to install the package. If those methods are registered in R<4 they cause failures for known reasons (which is the specific fix in R 4+).  So the `if()` must be there in NAMESPACE and hence the version of R used to install package must match major version with the version using the installed package. Only because this fix in R-devel happens to be in 4.0.0 so it was convenient to use the "major version matching" words. But for experts like you who know about historical binary incompatibilities in R versions and that R 4+ doesn't have anything like that, I can see how you thought data.table was wrong.

The message correctly uses the words 'installed using' ... 

```This is R 4.0.0 but data.table has been installed using R 3.6.1. The major version must match. Please reinstall data.table.```

To work in R 4+ data.table does need to be installed using R 4+.  You can use the same .zip that was built using R 3 though, it doesn't need to be rebuilt or recompiled.  The same binary package just needs to be reinstalled.

**Comment by gaborcsardi** [comment]
@mattdowle Oh, I got it. Right. So in this case it is indeed true that the 3.6.1 build is incompatible with R 4.0.0.

Which is unfortunate, because that's what CRAN provides under R-devel builds.

**Comment by gaborcsardi** [comment]
Btw. it is not enough to reinstall the zip, you need to rebuild it. Re-installing the zip is just equivalent to extacting it, and the NAMESPACE etc. does not run, no R code runs, AFAIK.

**Comment by mattdowle** [comment]
Hi @gaborcsardi. Thanks for looking into this.

> Btw. it is not enough to reinstall the zip, you need to rebuild it. Re-installing the zip is just equivalent to extacting it, and the NAMESPACE etc. does not run, no R code runs, AFAIK.

Is that a difference between Windows and Linux?  Because on Linux I don't need to rebuild.  R CMD INSTALL on Linux does appear to run the if() in NAMESPACE.  Did you try on Windows?

How would R-hub cope if there was a binary incompatibility in R 4 vs 3?  Like there was in R 3 vs 2, iirc.

**Comment by gaborcsardi** [comment]
> Is that a difference between Windows and Linux?

You mean a Linux _binary_ package? AFAIK there is no different between Windows and Linux for _binary_ packages. But on Linux usually you install a source package. I.e. if you do `R CMD INSTALL --build` on Linux, you get a binary package, and if you `R CMD INSTALL` that, it does not look at `NAMESPACE`.

> How would R-hub cope if there was a binary incompatibility in R 4 vs 3?

R's job is to cope with that. And it does, mostly. I.e. when you load a package, it checks if it is binary compatible with the currently running version. But apparently it does not check for code in the `NAMESPACE`. 

The `NAMESPACE` is parsed when you `R CMD INSTALL` the _source_ package, which is on R 3.6.1 for the CRAN binaries, and then when we load that on R 4.0.0. is just loads the already parsed namespace information from an `.rds` file.

So in summary, I think this is an R bug, that is probably worth reporting.

**Comment by mattdowle** [comment]
@gaborcsardi I just tried R-devel on Windows and this is what I get.  This seems right: Windows binaries are not being built yet for R 4 by CRAN it seems.  I don't see what the R bug is.
![image](https://user-images.githubusercontent.com/2779998/68238842-c118ae00-ffbe-11e9-8de3-612c8a08efb2.png)

**Comment by mattdowle** [comment]
I emailed Kurt and Uwe and copied you, @gaborcsardi.

**Comment by mattdowle** [comment]
@gaborcsardi  Following Uwe's email reply to us, I retested just now on WIndows. Indeed CRAN is now producing binaries for R 4 series and data.table loads and passes tests; i.e. https://cloud.r-project.org/bin/windows/contrib/4.0/ is available and working now.
Does this solve the issue on R-Hub?

**Comment by DyfanJones** [comment]
@mattdowle this has resolved the issue for me, many thanks for all your hard work.

**Comment by mattdowle** [comment]
@tdhock @gaborcsardi  This one can be closed now I think, iiuc.
